src/dev/HallSensor.cpp

```diff
@@ -0,0 +1,112 @@
+    const bool magnetDetected = gpio.readPin() == MAGNET_DETECTED_STATE;
+    bool magnetDetectedNow = false;
+    if (!magnetWasDetected && magnetDetected != magnetInLastRead && magnetDetected) {
+        magnetWasDetected = true;
+        magnetDetectedNow = true;
+    } else if (magnetWasDetected && magnetDetected == magnetInLastRead) {
+        magnetWasDetected = false;
+    }
+    magnetInLastRead = magnetDetected;
+
+    switch (state) {
+    case WheelSpeedState::STOP:
+        if (magnetDetectedNow) {
+            prevTime = now;
+            state = WheelSpeedState::INITIALIZING;
+        }
+        break;
+
+    case WheelSpeedState::INITIALIZING:
+        if (magnetDetectedNow) {
+            // If the magnet is detected for the second time, record the time and the first
+            // interval and enter the maintain state
+            lastInterval = elapsedTime;
+            prevTime = now;
+            state = WheelSpeedState::MAINTAIN;
+        }
+
+        // If the magnet hasn't been detected for a significant amount of time, go back to the stop
+        // state
+        if (elapsedTime > THRESHOLD) {
+            state = WheelSpeedState::STOP;
+            prevTime = 0;
+        }
+        break;
+
+    case WheelSpeedState::MAINTAIN:
+        if (magnetDetectedNow) {
+            // If the magnet is newly detected, record the time and interval
+            lastInterval = elapsedTime;
+            prevTime = now;
+        }
+
+        // If the magnet hasn't been detected for a significant amount of time, go back to the stop
+        // state
+        if (elapsedTime > THRESHOLD) {
+            state = WheelSpeedState::STOP;
+            prevTime = 0;
+            lastInterval = 0;
+            // If the elapsed time since the magnet was last detected exceeds the last interval, record
+            // the last interval as if the magnet was just detected because the bike is slowing
+            // down, but the exact speed can't be calculated
+        } else if (elapsedTime > lastInterval) {
+            lastInterval = elapsedTime;
+        }
+        break;
+    }
+}
+
+uint32_t HallSensor::getRawInterval() {
+    return lastInterval;
+}
+
+uint32_t HallSensor::getSpeed() {
+    if (state == WheelSpeedState::STOP) {
+        // The wheel isn't moving, so its speed is zero.
+        return 0;
+    }
+
+    /*
+     * The speed of the wheel in revolutions per minute.
+     * lastInterval is the time that it takes for one full revolution of the wheel
+     * in milliseconds, so converting that to RPM is just dividing 60000 by lastInterval.
+     */
+    const uint32_t rpm = 60000 / lastInterval;
+    /*
+     * The speed of the bike in miles per hour.
+     * RPM * (circumference in inches (2 * pi * wheelRadius) / 1 rotation) * (1 mile / 63360 inches) * (60 minutes / hour)
+     * is what calculates the speed in mph.
+     */
+    const uint32_t speed = (rpm * ((wheelRadius * 2 * 3142) / numberOfMagnets) * 60) / 63360000;
```

Comment:
The value of `numberOfMagnets` should affect the value of `rpm`. This equation might still work, but if someone tries to debug it, they're going to get a nonsense value when they print `rpm`